examples/vision/3D_image_classification.py

```diff
@@ -0,0 +1,469 @@
+
+"""
+## Build train and test datasets
+Read the scans from the class directories and assign labels.
+Lastly, split the dataset into train and test subsets.
+"""
+
+# read and process the scans
+# each scan is resized across width, height, and depth
+right_scans = np.array([process_scan(path) for path in right_scan_paths])
+left_scans = np.array([process_scan(path) for path in left_scan_paths])
+
+# for the CT scans having cancerous regions in the right side
+# assign 1, similarly for left assign 0.
+right_labels = np.array([1 for _ in range(len(right_scans))])
+left_labels = np.array([0 for _ in range(len(left_scans))])
+
+# split data in the ratio 70-30 for training and testing
+x_train = np.concatenate((right_scans[:14], left_scans[:14]), axis=0)
+y_train = np.concatenate((right_labels[:14], left_labels[:14]), axis=0)
+x_test = np.concatenate((right_scans[14:], left_scans[14:]), axis=0)
+y_test = np.concatenate((right_labels[14:], left_labels[14:]), axis=0)
+print(
+    "Number of samples in train and test are %d and %d."
+    % (x_train.shape[0], x_test.shape[0])
+)
+
+"""
+## Preprocessing and data augmentation
+CT scans store raw voxel intensity in Hounsfield units (HU). They range from
+-1024 to above 2000 in this dataset. Above 400 are bones with different
+radiointensity, so this is used as a higher bound. A threshold between
+-1000 and 400 are commonly used to normalize CT scans. The CT scans are
+also augmented by rotating and blurring. There are different kinds of
+preprocessing and augmentation techniques out there, this example shows a few
+simple ones to get started.
+"""
+
+import random
+
+from scipy import ndimage
+from scipy.ndimage import gaussian_filter
+
+
+def normalize(volume):
+    min = -1000
+    max = 400
+    volume = (volume - min) / (max - min)
+    # rescale between 0-1
+    volume[volume > 1] = 1.0
+    volume[volume < 0] = 0.0
+    volume = np.expand_dims(volume, axis=3)
+    return volume
+
+
+@tf.function(input_signature=[tf.TensorSpec((128, 128, 128), tf.float64)])
+def tf_normalize(volume):
+    normalized_volume = tf.numpy_function(normalize, [volume], tf.float64)
+    return normalized_volume
+
+
+def rotate(volume):
+    # define some rotation angles
+    angles = [-20, -10, -5, 5, 10, 20]
+    # pick angles at random
+    angle = random.choice(angles)
+    # rotate volume
+    volume = ndimage.rotate(volume, angle, reshape=False)
+    return volume
+
+
+@tf.function(input_signature=[tf.TensorSpec((128, 128, 128), tf.float64)])
+def tf_rotate(volume):
+    augmented_volume = tf.numpy_function(rotate, [volume], tf.float64)
+    return augmented_volume
+
+
+def blur(volume):
+    # gaussian blur
+    volume = gaussian_filter(volume, sigma=1)
+    return volume
+
+
+@tf.function(input_signature=[tf.TensorSpec((128, 128, 128), tf.float64)])
+def tf_blur(volume):
+    augmented_volume = tf.numpy_function(blur, [volume], tf.float64)
+    return augmented_volume
+
+
+def train_augment(volume, label):
```

Comment:
Let's simplify these 2 functions:

- normalization should be done with TF ops, not numpy. Only rotation and blur require scipy.
- no need for conditionals
- It seems very restrictive to hardcode image dimensions in the tensorspec, remove them.
- the scipy functions can be inlined inside their tf wrapper